### ISSUE TITLE
Assignment expression should be typed as RHS rather than LHS in templates

### ISSUE THREAD

**Issue by RedBeard0531** [microsoft/TypeScript#49877]
# Bug Report

### 🔎 Search Terms

(All search terms are in title)

### 🕗 Version & Regression Information

- This changed between versions 4.2.3 and 4.3.5

It still repos with 4.7.4, 4.8-beta and the current nightly (4.8.0-dev.20220712).

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.4#code/C4TwDgpgBAKu0F4oGdgCcCWA7A5lAPlFgK4C2ARhGgFDUA2EwUA9uQFYBcUAShAMbM0AEwA8qTLgA0seAD4oSAN4BfagDNiWPsAzMsUYBFQiYUCAA9DWIchmRZACgBuAQzpcYASg9RF1KFAYalAOAAyevv4BUAD0MbAAFhi2oJB8CfwA1rbkxEzJUADugkIYVAYJLvpYEBBCdQB0UQGsbADaAOSGqB0AugpQrnQA3M1QaIzEaPpDowGqUXFQAHIYfNBVQlACWMjJENK5TGouGHQp8NsZfJnYOE0BE8BT+q2d3cB9A7PUCwxM4juAw+Dg6GTodGYHQiS2QCWYxDoW0oUDU2AgQA)

### 💻 Code

```ts
type Type = string | number

let obj: Record<string, Type> = {}
function test<T extends Type>(val: T): T {
  if (0) {
    // This typechecks but is wordier than needed.
    obj['test'] = val;
    return val;
  }

  // Nice and concise, but fails type checking.
  return obj['test'] = val;
}

let string = test('hello') // should be fine
```

### 🙁 Actual behavior

Fails typechecking. It appears to be because the expression `obj['test'] = val` is given the type of the broader LHS (`Type`) rather than the narrower type of the RHS (`T`). 

### 🙂 Expected behavior

This should typecheck correctly. Given that the RHS of an assignment always needs to be more narrow than the LHS, it seems like it makes the most sense to use the RHS. One possible exception would be to fall back to the LHS if the RHS is `any`.

This would also be nice for consistency with top-level code where it [seems to work correctly](https://www.typescriptlang.org/play?ts=4.7.4#code/C4TwDgpgBAKu0F4oGdgCcCWA7A5lAPlFgK4C2ARhGgFDUA2EwUA9uQFYBcUAShAMbM0AEwA8qTLgA0seAD4oSAN4BfWgyYAPBS3YBtAOTAIqfQF1t+gBYZ9AbmoB6B1CgA9APy0BWVFBDbWNgMjE3MkKxt7JxcXD1pqIA)

**Comment by RyanCavanaugh** [comment]
> This should typecheck correctly. Given that the RHS of an assignment always needs to be more narrow than the LHS

The underlying logic here is correct but the problem is that this is often *too* specific. For example:
```ts
function fn() {
  let s: string[];
  return s = [];
}
```
`fn`'s return type is `string[]` here because we're using the LHS type. If we used the RHS type it'd be `never[]`, which is not a type you really want manifested in a program.

**Comment by RedBeard0531** [comment]
I don't think it is working as intended, since if I manually instantiate the template by substituting `string` for `T` [it _does_ typecheck](https://www.typescriptlang.org/play?ts=4.7.4#code/C4TwDgpgBAKu0F4oGdgCcCWA7A5lAPlFgK4C2ARhGgFDUA2EwUA9uQFYBcUAShAMbM0AEwA8qTLgA0seAD4oSAN4BfagDNiWPsAzMsUYBFQAKAG4BDOl3HYcASmvpbURdShQMaqMYAMdl27uUAD0wbAAFhjIBvB84fwA1tHkxExRUADugkIYVAbh5vpYEBBCpQB0ge6sbADaAOSGqPUAugpQFnQA3FVQaIzEaPqdPe6qgaFQAHIYfNCFQlACWMhRENIpTGrmGHTRoJBL8XwJtpXu-cCD+jUNTcCt7SPU4wxMNrjt98b18XR0zHq-kmyHCzGIdEWlCgamwECAA).

I [did some more digging](https://www.typescriptlang.org/play?ts=4.7.4#code/C4TwDgpgBAMgFgZwCrmldGMF4oOAJwEsA7AcygB8piBXAWwCMJ9LqaAbdgKFEigHVCAEwhIIdMOwCGwUaig48RMq1qNmXLuwjAoAewYArAFxQAShADGe-EIA8SkqQA0sRCkgA+BVADeAXy4AMxpiS2BCPWIoWTw7JCgIAA9ZYiEEAWFRcUkZOS8ACgA3KXZTJABKcr8udEIgqAKABgqazAB6dqgkOEIM3is4KwBrDIYaXT6oAHcbIUJmGLgpaOIICBEhADpajANDAG0AcljgI4BdHxL2AG5d9HwdGnxo67v0QN3OqAA5QktoCshFBrMQEH0IK5xroglJCOx+vJLENLMMnDtdo9gM9ovtjqcLldSnddtpdAg4Hpph5oDg8ScIHhCThrl8ulAAHoAfjZ6AAypSOMCmN0oNCoFNBCIxBJpLIaRKwbIpEIuJ8ybgCE4fKcCkchpw9EdWt8KXohWLoEESBAgA) and I realized I was wrong about the cause of the bug. It _isn't_ that it is using the LHS of the assignment. The problem is that it is using `WideTemplateType` in `func<T extends WideTemplateType>` rather than the precise type `T` (not sure what the technical names for those things are in TS). Here is a better code example, same as the last playground link:

```ts
type LhsType          = string | number | null
type WideTemplateType = string | number

let obj: Record<string, LhsType> = {}
function test<T extends WideTemplateType>(val: T): T {
  if (0) {
    // This typechecks but is wordier than needed.
    obj['test'] = val;
    return val;
  }

  // Nice and consise, but fails type checking.
  return obj['test'] = val;

  let showType = obj['test'] = val
  //  ^?
  //  Should be T but is WideTemplateType instead
}

let string = test('hello') // should be fine
```

Should I A) update this ticket's description and summary to reflect this new information, B) close this ticket and open a new clean one, or C) do neither?

**Comment by RyanCavanaugh** [comment]
That's kind of weird.

**Comment by fatcerberus** [comment]
> The problem is that it is using `WideTemplateType` in `func<T extends WideTemplateType>` rather than the precise type `T` (not sure what the technical names for those things are in TS).

For future reference: `T` is a *type parameter* and `WideTemplateType` is its *constraint*.
